firestore/integration_test_internal/src/leveldb_snappy_test.cc

```diff
@@ -49,36 +49,109 @@ using ::firebase::firestore::util::Path;
+// callback is invoked with a `status` where `status.ok()` is not true, then
+// iteration will stop and the callback will not be invoked again.
+void IterateOverLevelDbDatabaseThatUsesSnappyCompression(
+    std::function<void(const leveldb::Status&)>);
 
-  Path leveldb_path = CreateLevelDbDatabaseThatUsesSnappyCompression();
-  if (HasFatalFailure()) return;
+#if FIREBASE_TESTS_BUILT_BY_CMAKE
 
-  leveldb::Options options;
-  options.create_if_missing = false;
+// Ensure that LevelDb is compiled with Snappy compression support.
+// See https://github.com/firebase/firebase-ios-sdk/pull/9596 for details.
+TEST(LevelDbSnappy, LevelDbSupportsSnappy) {
+  IterateOverLevelDbDatabaseThatUsesSnappyCompression(
+      [](const leveldb::Status& status) {
+        ASSERT_TRUE(status.ok()) << ConvertStatus(status);
+      });
+}
+
+#else  // FIREBASE_TESTS_BUILT_BY_CMAKE
+
+// Ensure that LevelDb is NOT compiled with Snappy compression support.
+TEST(LevelDbSnappy, LevelDbDoesNotSupportSnappy) {
+  bool got_failed_status = false;
+  IterateOverLevelDbDatabaseThatUsesSnappyCompression(
+      [&](const leveldb::Status& status) {
+        if (!status.ok()) {
+          got_failed_status = true;
+          ASSERT_TRUE(status.IsCorruption()) << ConvertStatus(status);
+        }
+      });
 
+  if (!HasFailure()) {
+    ASSERT_TRUE(got_failed_status)
+        << "Reading a Snappy-compressed LevelDb database was successful; "
+           "however, it should NOT have been successful "
+           "since Snappy support is expected to NOT be available.";
+  }
+}
+
+#endif  // FIREBASE_TESTS_BUILT_BY_CMAKE
+
+void IterateOverLevelDbDatabaseThatUsesSnappyCompression(
+    std::function<void(const leveldb::Status&)> callback) {
   std::unique_ptr<leveldb::DB> db;
   {
+    Path leveldb_path = CreateLevelDbDatabaseThatUsesSnappyCompression();
+    if (leveldb_path.empty()) {
+      return;
+    }
+
+    leveldb::Options options;
+    options.create_if_missing = false;
+
     leveldb::DB* db_ptr;
     leveldb::Status status =
         leveldb::DB::Open(options, leveldb_path.ToUtf8String(), &db_ptr);
-    ASSERT_TRUE(status.ok());
+
+    ASSERT_TRUE(status.ok())
+        << "Opening LevelDb database " << leveldb_path.ToUtf8String()
+        << " failed: " << ConvertStatus(status);
+
     db.reset(db_ptr);
   }
 
-  // One of the assertions below will fail when LevelDb attempts to read a block
-  // that is compressed with Snappy and Snappy compression support is not
-  // compiled in.
   std::unique_ptr<leveldb::Iterator> it(
       db->NewIterator(leveldb::ReadOptions()));
   for (it->SeekToFirst(); it->Valid(); it->Next()) {
-    ASSERT_TRUE(it->status().ok()) << ConvertStatus(it->status());
+    callback(it->status());
+    if (!it->status().ok()) {
+      return;
+    }
+  }
+
+  // Invoke the callback on the final status.
+  callback(it->status());
+}
+
+template <typename T>
+void WriteFile(const Path& dir,
```

Comment:
Note to reviewer(s): The `WriteFile()` method was simply moved to later on in the file; its contents were not changed.